### PR TITLE
More examples in the backend

### PR DESCRIPTION
**Pull request by sifnoc**
Related issue: https://github.com/summa-dev/summa-solvency/issues/128

Major changes in the Backend:

- By default, the inclusion proof in the Round is now generated using the `gen_evm_proof_shplonk` method.
- Introduced `full_evm_verifier` to verify inclusion proofs created with `gen_evm_proof_shplonk`.

### CODE REVIEW

File: backend/README.md [hunk 1]
```diff
@@ -65,25 +65,140 @@ By completing these steps, the backend will be primed with the essential verifie
...
+To initialize the Round, several parameters are required, including paths to specific CSV files (`assets.csv` and `entry_16.csv`), as well as a path to the ptau file (`ptau/hermez-raw-11`).
+
+The roles of these files are as follows:
+- `assets.csv`: This file is essential for calculating the total balance of assets for the solvency proof. Currently, only the CEX can generate this asset CSV file in its specific manner.
+
+- `entry_16.csv`: This file is used to build the Merkle sum tree, where each leaf element originates from sixteen entries in the CSV.
+X
+- `ptau/hermez-raw-11`: Contains the Powers of Tau trusted setup parameters, essential for constructing the zk circuits.
+
+By employing the round instance, the solvency proof is dispatched to the Summa contract using the `dispatch_solvency_proof` method.
```
Comment (alxkzmn) [inline backend/README.md:132]: "An instance of `Round` dispatches the solvency proof using the `dispatch_solvency_proof` method."

File: backend/README.md [hunk 2]
```diff
@@ -65,25 +65,140 @@ By completing these steps, the backend will be primed with the essential verifie
...
+
+To excute this example:
```
Comment (alxkzmn) [inline backend/README.md:134]: execute

File: backend/README.md [hunk 3]
```diff
@@ -65,25 +65,140 @@ By completing these steps, the backend will be primed with the essential verifie
...
-On successful execution, you'll observe a message indicating the verification outcome:
+### 4. Generating and Exporting Inclusion Proofs
+
+Assuming you are a CEX, let's say you've already committed the `solvency` and `ownership` proofs to the Summa contract. Now, you need to generate inclusion proofs for every user.
+
+In this example, we demonstrate how to generate and export user-specific inclusion proofs using the Round. This proof is crucial for users as it helps them validate the presence of specific elements within the Merkle sum tree, which forms a part of the solvency proof submitted.
+
+After generating the inclusion proof, end of example parts the inclusion proof is transformed into a JSON format, making it easily shareable.
+
+To excute this example:
```
Comment (alxkzmn) [inline backend/README.md:153]: Can probably be fixed everywhere at once by running find&replace

File: backend/examples/generate_signatures.rs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+    // Generate signatures for the given 'message' using the mock signer.
+    // For this example, the 'mock_signer' file contains only the 'sign_message' function.
+    // CEX should implement their own signer and use it here instead of 'sign_message'.
+    let signatures = sign_message(message).await?;
+
+    // Write the signatures to a CSV file to be used in the `verify_signatures` example.
+    // It's envisioned that this CSV file will remain internal to CEX; only the Summa contract will publish its contents.
+    let file = File::create(path)?;
+    let mut wtr = WriterBuilder::new().delimiter(b';').from_writer(file);
```
Comment (alxkzmn) [inline backend/examples/generate_signatures.rs:24]: Why isn't a comma sufficient?

**Review by alxkzmn — CHANGES_REQUESTED** [review]
I think it's not necessary to use a ";" as a delimiter in the CSV.

File: backend/examples/generate_signatures.rs (hunk 4 above)
Comment (sifnoc) [reply backend/examples/generate_signatures.rs:24]: Yeah, It's sufficient. I will fix this too.

Comment (sifnoc) [reply backend/examples/generate_signatures.rs:24]: Fixed at [eacfe7d](https://github.com/summa-dev/summa-solvency/pull/159/commits/eacfe7df0542f35ab9367cc34ad8a298b0cdda5d)

File: backend/README.md (hunk 3 above)
Comment (sifnoc) [reply backend/README.md]: Fixed at [eacfe7d](https://github.com/summa-dev/summa-solvency/pull/159/commits/eacfe7df0542f35ab9367cc34ad8a298b0cdda5d)

File: backend/README.md (hunk 2 above)
Comment (sifnoc) [reply backend/README.md]: Fixed at [eacfe7d](https://github.com/summa-dev/summa-solvency/pull/159/commits/eacfe7df0542f35ab9367cc34ad8a298b0cdda5d)

File: backend/README.md (hunk 1 above)
Comment (sifnoc) [reply backend/README.md]: Replaced at [eacfe7d](https://github.com/summa-dev/summa-solvency/pull/159/commits/eacfe7df0542f35ab9367cc34ad8a298b0cdda5d)

File: backend/README.md [hunk 5]
```diff
@@ -65,25 +65,140 @@ By completing these steps, the backend will be primed with the essential verifie
...
+
+Key points to note:
+
+The instance of `AddressOwnership` is initialized with `signatures.csv`, and is named `address_ownership_client`. This instance has already loaded the signature data.
+
+The `dispatch_proof_of_address_ownership` function sends a transaction to the Summa contract, registering the addresses owned by the CEX on the contract.
+
+After dispatching the transaction via `dispatch_proof_of_address_ownerhip`, the example computes the hashed addresses (address_hashes) to verify they have been correctly registered on the Summa contract.
+
+To excute this example:
```
[the remark this answers is not in the document]
Comment (sifnoc) [reply backend/README.md]: Fixed at [eacfe7d](https://github.com/summa-dev/summa-solvency/pull/159/commits/eacfe7df0542f35ab9367cc34ad8a298b0cdda5d)

File: backend/README.md [hunk 6]
```diff
@@ -65,25 +65,140 @@ By completing these steps, the backend will be primed with the essential verifie
 
 ## Examples
 
-### Running the Inclusion Verification
+The sequence in which the examples are introduced closely relates to the steps of the Summa protocol.
+These examples will be help to understand how the Summa works with the Summa contract and the user side.
```
[the remark this answers is not in the document]
Comment (sifnoc) [reply backend/README.md]: Fixed at [eacfe7d](https://github.com/summa-dev/summa-solvency/pull/159/commits/eacfe7df0542f35ab9367cc34ad8a298b0cdda5d)

File: backend/examples/verify_inclusion_on_local.rs [hunk 7]
```diff
@@ -0,0 +1,61 @@
+#![feature(generic_const_exprs)]
```
Comment (enricobottazzi) [inline backend/examples/verify_inclusion_on_local.rs:1]: I suggest to remove `verify_inclusion_on_local` from the examples and keep `verify_inclusion_on_contract` as the verification method by default

**Review by enricobottazzi — CHANGES_REQUESTED** [review]

File: zk_prover/src/circuits/utils.rs [hunk 8]
```diff
@@ -134,6 +134,43 @@ pub fn full_verifier(
...
+    verify_proof::<_, VerifierSHPLONK<'_, Bn256>, _, EvmTranscript<_, _, _, _>, _>(
+        params,
+        vk,
+        AccumulatorStrategy::new(params.verifier_params()),
+        instances,
+        &mut transcript,
+    )
+    .is_ok()
+
+    // let accept = {
```
[the remark this answers is not in the document]
Comment (sifnoc) [reply zk_prover/src/circuits/utils.rs:157]: Yes! I forgot to remove.
I will remove in next commit

File: backend/examples/helpers/inclusion_proof.rs [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+use summa_solvency::merkle_sum_tree::Entry;
+
+#[derive(Debug, Deserialize)]
+pub struct InclusionProof {
+    pub leaf_hash: String,
+    pub root_hash: String,
+    pub proof: String,
+}
+
+pub fn generate_leaf_hash<const N_ASSETS: usize>(user_name: String, balances: Vec<usize>) -> Fp
```
[the remark this answers is not in the document]
Comment (sifnoc) [reply backend/examples/helpers/inclusion_proof.rs:14]: Sure! I would love to

File: backend/examples/helpers/inclusion_proof.rs [hunk 10]
```diff
@@ -0,0 +1,25 @@
+use halo2_proofs::halo2curves::bn256::Fr as Fp;
+use num_bigint::BigUint;
+use serde::Deserialize;
+
+use summa_solvency::merkle_sum_tree::Entry;
+
+#[derive(Debug, Deserialize)]
+pub struct InclusionProof {
```
[the remark this answers is not in the document]
Comment (sifnoc) [reply backend/examples/helpers/inclusion_proof.rs:8]: Yes, Indeed, but I have to check `MstInclusionProof` can be used in parsing json.

File: backend/src/tests.rs [hunk 11]
```diff
@@ -226,27 +238,23 @@ mod test {
...
-
-        let public_inputs = vec![
-            U256::from_big_endian(&leaf_hash),
-            U256::from_big_endian(&root_hash),
-        ];
-
-        // Ensure the root hash matches the one from the contract
-        let onchain_mstroot = summa_contract.mst_roots(U256::from(1)).await.unwrap();
-        assert_eq!(onchain_mstroot, U256::from_big_endian(&root_hash));
+        let public_inputs: Vec<U256> = inclusion_proof
```
[the remark this answers is not in the document]
Comment (sifnoc) [reply backend/src/tests.rs:241]: Odd, I'm not receiving any such warning

File: backend/examples/mock_signer.rs [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+use ethers::{
+    abi::{encode, Token},
+    signers::{LocalWallet, Signer},
+    utils::{keccak256, to_checksum},
+};
+
+use summa_backend::apis::csv_parser::SignatureRecord;
+
+// Separated this function from the `generate_signatures.rs` for clarity on the example.
+pub async fn sign_message(message: &str) -> Result<Vec<SignatureRecord>, Box<dyn Error>> {
```
[the remark this answers is not in the document]
Comment (sifnoc) [reply backend/examples/mock_signer.rs]: Removed at [c9643c9](https://github.com/summa-dev/summa-solvency/pull/159/commits/c9643c948ebcdae61ba46f72fde841acb6710233)

File: backend/examples/generate_signatures.rs [hunk 13]
```diff
@@ -0,0 +1,34 @@
+#![feature(generic_const_exprs)]
+use std::{error::Error, fs::File};
+
+use csv::WriterBuilder;
+
+mod mock_signer;
+use mock_signer::sign_message;
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn Error>> {
```
[the remark this answers is not in the document]
Comment (sifnoc) [reply backend/examples/generate_signatures.rs]: Also Removed at [c9643c9](https://github.com/summa-dev/summa-solvency/pull/159/commits/c9643c948ebcdae61ba46f72fde841acb6710233)

File: backend/README.md [hunk 14]
```diff
@@ -65,25 +65,140 @@ By completing these steps, the backend will be primed with the essential verifie
...
+Upon successful execution, you should see the message:
+```
+Ownership proofs are submitted successfully!
+```
+
+Reminder: This demonstration takes place in a test environment. In real-world production, always ensure that the Summa contract is correctly deployed on the target chain.
+
+### 3. Submit Proof of Solvency
+
+Before generate inclusion proof for every user of the current round, You should submit proof of sovlency to Summa contract. Currently, we made this as mandatory way to commit the root hash of the Merkle Sum Tree.
```
[the remark this answers is not in the document]
Comment (sifnoc) [reply backend/README.md]: Fixed at [c9643c9](https://github.com/summa-dev/summa-solvency/pull/159/commits/c9643c948ebcdae61ba46f72fde841acb6710233)

File: backend/README.md [hunk 15]
```diff
@@ -65,25 +65,140 @@ By completing these steps, the backend will be primed with the essential verifie
...
+
+- `entry_16.csv`: This file is used to build the Merkle sum tree, where each leaf element originates from sixteen entries in the CSV.
+X
+- `ptau/hermez-raw-11`: Contains the Powers of Tau trusted setup parameters, essential for constructing the zk circuits.
+
+An instance of Round dispatches the solvency proof using the `dispatch_solvency_proof` method.
+
+To execute this example:
+```
+cargo run --example submit_solvency
```
[the remark this answers is not in the document]
Comment (sifnoc) [reply backend/README.md]: Updated at [c9643c9](https://github.com/summa-dev/summa-solvency/pull/159/commits/c9643c948ebcdae61ba46f72fde841acb6710233)

File: backend/README.md [hunk 16]
```diff
@@ -65,25 +65,140 @@ By completing these steps, the backend will be primed with the essential verifie
...
+
+```
+ "Exported proof to user #0, as `user_0_proof.json`"
+```
+
+### 5. Verify Proof of Inclusion
+
+This is the final step in the Summa process and the only part that occurs on the user side.
+
+The user will receive the proof for a specific Round. There are two ways to verify the proof, one is on binary verifier in local environment, another is that the verifier function on the Summa contract.
```
[the remark this answers is not in the document]
Comment (sifnoc) [reply backend/README.md]: Updated at [c9643c9](https://github.com/summa-dev/summa-solvency/pull/159/commits/c9643c948ebcdae61ba46f72fde841acb6710233)

File: backend/README.md [hunk 17]
```diff
@@ -65,25 +65,140 @@ By completing these steps, the backend will be primed with the essential verifie
...
+On-chain Function Verification: The user then invokes the `verify_inclusion_proof` method on the Summa contract. Since this is a view function, it returns a boolean value without incurring any gas fees, indicating the success or failure of the verification.
+
+To run the verify inclusion on contract example:
+```
+cargo run --example verify_inclusion_on_contract
+```
+
+You will see the result like:
+```
+Verifying the proof on contract veirifer for User #0: true
```
[the remark this answers is not in the document]
Comment (sifnoc) [reply backend/README.md]: Fixed at  [c9643c9](https://github.com/summa-dev/summa-solvency/pull/159/commits/c9643c948ebcdae61ba46f72fde841acb6710233)

File: backend/README.md [hunk 18]
```diff
@@ -65,25 +65,140 @@ By completing these steps, the backend will be primed with the essential verifie
 
 ## Examples
```
[the remark this answers is not in the document]
Comment (sifnoc) [reply backend/README.md]: Combined add [5c9bf49](https://github.com/summa-dev/summa-solvency/pull/159/commits/5c9bf49d01ceb21e40d72b56fd4403ec99059f4c)

File: backend/src/tests.rs [hunk 19]
```diff
@@ -226,27 +238,23 @@ mod test {
...
+            })
+            .collect();
 
         // Verify inclusion proof with onchain function
-        let _result = summa_contract
+        summa_contract
             .verify_inclusion_proof(proof, public_inputs, U256::from(1))
-            .await;
+            .await
+            .unwrap();
```
[the remark this answers is not in the document]
Comment (sifnoc) [reply backend/src/tests.rs]: Sure, Added assertion at [c9643c9](https://github.com/summa-dev/summa-solvency/pull/159/commits/c9643c948ebcdae61ba46f72fde841acb6710233)

File: backend/src/tests.rs [hunk 20]
```diff
@@ -81,52 +129,16 @@ mod test {
     };
 
     use crate::apis::{address_ownership::AddressOwnership, round::Round};
```
[the remark this answers is not in the document]
Comment (sifnoc) [reply backend/src/tests.rs:111]: Removed at  [c9643c9](https://github.com/summa-dev/summa-solvency/pull/159/commits/c9643c948ebcdae61ba46f72fde841acb6710233)

File: zk_prover/src/circuits/utils.rs [hunk 21]
```diff
@@ -134,6 +134,43 @@ pub fn full_verifier(
     .is_ok()
 }
 
+/// Verifies a proof that generated by the `gen_evm_proof_shplonk` function.
+pub fn full_evm_verifier(
```
[the remark this answers is not in the document]
Comment (sifnoc) [reply zk_prover/src/circuits/utils.rs]: Removed at [5c9bf49](https://github.com/summa-dev/summa-solvency/pull/159/commits/5c9bf49d01ceb21e40d72b56fd4403ec99059f4c)

File: backend/examples/helpers/inclusion_proof.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (sifnoc) [reply backend/examples/helpers/inclusion_proof.rs]: Removed at [5c9bf49](https://github.com/summa-dev/summa-solvency/pull/159/commits/5c9bf49d01ceb21e40d72b56fd4403ec99059f4c)

File: backend/examples/verify_inclusion_on_local.rs (hunk 7 above)
Comment (sifnoc) [reply backend/examples/verify_inclusion_on_local.rs]: It make sense, removed at [c9643c9](https://github.com/summa-dev/summa-solvency/pull/159/commits/c9643c948ebcdae61ba46f72fde841acb6710233)

File: backend/README.md [hunk 22]
```diff
@@ -63,27 +63,86 @@ The following steps are optional and are only required if you need to update the
...
 
-To run the example:
 ```
-cargo run --example verify_inclusion
+1. Ownership proofs are submitted successfully!
 ```
 
-On successful execution, you'll observe a message indicating the verification outcome:
+
+### 2. Submit Proof of Solvency
```
Comment (enricobottazzi) [inline backend/README.md:98]: Can you specify that the goal of the proof of solvency is also to check that the assets controlled by the exchange are greater than the liabilities (as committed in the Merkle Sum Tree Root Hash)

**Review by enricobottazzi — APPROVED** [review]

**Review by alxkzmn — APPROVED** [review]

**Pull request merged**